Added guidance on field tags in marshalled structs (#148)
uber-go/guide: style.md

</td></tr>
</tbody></table>

### Use field tags in marshalled structs

> Any struct field that is marshalled into JSON, YAML or other formats, must have relevant tag.

Reviewer: nit: move the last comma, add a "the":

> Any struct field that is marshalled into JSON, YAML, or other formats must have the relevant tag.

Alternatively, we should probably clarify "formats that support tag-based field naming" (not all do):

```suggestion
Any struct field that is marshaled into JSON, YAML, or other formats that support tag-based field naming should be annotated with the relevant tag.
```

> Any struct field that is marshaled into JSON, YAML, or other formats that support tag-based field naming should be annotated with the relevant tag.

Reviewer 2: line breaks here would be nice:

```suggestion
Any struct field that is marshaled into JSON, YAML,
or other formats that support tag-based field naming
should be annotated with the relevant tag.
```

<table>
<thead><tr><th>Bad</th><th>Good</th></tr></thead>
<tbody>
<tr><td>

```go
type T struct {
  Number int
  String string
}
bytes, err := json.Marshal(T{
  Number: 137,
  String: "uber",
})
```

</td><td>

```go
type T struct {
  Number int    `json:"number"`
  String string `json:"string"`
}
bytes, err := json.Marshal(T{
  Number: 137,
  String: "uber",
})

> ```

Author: good call, updated the examples

</td></tr>
</tbody></table>

### Use field tags in marshaled structs

Any struct field that is marshaled into JSON, YAML, or other formats that support tag-based field naming should be annotated with the relevant tag.

<table>
<thead><tr><th>Bad</th><th>Good</th></tr></thead>
<tbody>
<tr><td>

```go
type Stock struct {
  Price int
  Name string
}
bytes, err := json.Marshal(Stock{
  Price: 137,
  Name: "UBER",
})
```

</td><td>

```go
type Stock struct {
  Price int    `json:"price"`
  Name string `json:"name"` // now it is safer to rename this field from Name to Symbol
}
bytes, err := json.Marshal(Stock{
  Price: 137,
  Name: "UBER",
})
```

</td></tr>
</tbody></table>

> The structure of JSON, YAML, and other formats, including field names, is a
> contract. In order to make that contract more explicit, field name struct tags
> should be used to ensure that fields names are consistent and do not change
> unexpectedly when changes to the structs fields are made.

Reviewer 2: I might reword this a bit.

```suggestion
Rationale:
The serialized form of the structure is a contract between different systems.
Changes to the structure of the serialized form--including field names--break
this contract. Specifying field names inside tags makes the contract explicit,
and it guards against accidentally breaking the contract by refactoring or
renaming fields.
```

</td></tr>
</tbody></table>

### Use field tags in marshaled structs

Any struct field that is marshaled into JSON, YAML,
or other formats that support tag-based field naming
should be annotated with the relevant tag.

<table>
<thead><tr><th>Bad</th><th>Good</th></tr></thead>
<tbody>
<tr><td>

```go
type Stock struct {

>   Price int
>   Name string
> }
> bytes, err := json.Marshal(Stock{
>   Price: 137,
>   Name: "UBER",

Reviewer 2: minor: alignment

```suggestion
  Price int
  Name  string
}

bytes, err := json.Marshal(Stock{
  Price: 137,
  Name:  "UBER",
```

})
```

</td><td>

```go

> type Stock struct {
>   Price int    `json:"price"`
>   Name string `json:"name"` // now it is safer to rename this field from Name to Symbol
> }
> bytes, err := json.Marshal(Stock{
>   Price: 137,
>   Name: "UBER",
> })

Reviewer 2: ```suggestion
type Stock struct {
  Price int    `json:"price"`
  Name  string `json:"name"`
  // Safe to rename Name to Symbol.
}

bytes, err := json.Marshal(Stock{
  Price: 137,
  Name:  "UBER",
})
```
